File: leetcode3/변지협/v2/3833. Count Dominant Indices.py
'''
1. 아이디어 :
문제에 명시되어있는 것 그대로 구현한다.

2. 시간복잡도 :
o(n^2)

3. 자료구조/알고리즘 :
'''
class Solution:
    def dominantIndices(self, nums: List[int]) -> int:
        n = len(nums)
        ans = 0
        for i in range(n-1):
            _sum = 0
            _len = n - i - 1
            print(i, _len)
            for j in range(i+1,n):
                _sum += nums[j]
            
            avg = _sum / _len
            
            if avg < nums[i]:
                ans +=1
        
        return ans

File: leetcode3/이진희/3833. Count Dominant Indices.java
/*

1. 아이디어 : 각 인덱스 기준 i+1부터 배열끝까지 합을 누적합으로 계산
              누적된 값의 평균이 배열 i보다 더 작다면 값 증가

2. 시간복잡도 : O(2N)

3. 자료구조/알고리즘 : 누적합

 */

class Solution {
    public int dominantIndices(int[] nums) {
        int[] cnt = new int[nums.length];
        int ans = 0;

        // len-2 ~ 0까지
        for(int i=nums.length-2; i>=0; i--) cnt[i]= cnt[i+1] + nums[i+1];

        for(int i=0; i<nums.length-1; i++) {
            double avg = (double)cnt[i]/(nums.length-1-i);
            if(nums[i]>avg) ans++;
        }

        return ans;
    }
}